client/dns_client.ml

```diff
@@ -1,87 +1,212 @@
+
+  let parse_response (type requested)
+    : requested Rr_map.key query_state -> Cstruct.t ->
+      [`Ok of requested | `Partial | `Msg of string] =
+    fun state buf ->
+    match consume_protocol_prefix buf state.protocol with
+    | Ok buf -> consume_rest_of_buffer state buf
+    | Error () -> `Partial
+
+end
+
+
+let stdlib_random n =
+  let b = Cstruct.create n in
+  for i = 0 to pred n do
+    Cstruct.set_uint8 b i (Random.int 256)
+  done;
+  b
+
+module type S = sig
+  type flow
+  type +'a io
+  type io_addr
+  type ns_addr = ([`TCP | `UDP]) * io_addr
+  type stack
+  type t
+
+  val create : ?rng:(int -> Cstruct.t) -> ?nameserver:ns_addr -> stack -> t
+
+  val nameserver : t -> ns_addr
+  val rng : t -> (int -> Cstruct.t)
+
+  val connect : ?nameserver:ns_addr -> t -> (flow, [> `Msg of string ]) result io
+  val send : flow -> Cstruct.t -> (unit, [> `Msg of string ]) result io
+  val recv : flow -> (Cstruct.t, [> `Msg of string ]) result io
+  val close : flow -> unit io
+
+  val bind : 'a io -> ('a -> 'b io) -> 'b io
+  val lift : 'a -> 'a io
+end
+
+module Make = functor (Transport:S) ->
+struct
+
+  type t = {
+    mutable cache : Dns_cache.t ;
+    clock : unit -> int64 ;
+    transport : Transport.t ;
+  }
+
+  let create ?(size=1000) ?rng ?nameserver ~clock stack =
+    { cache = Dns_cache.empty size ;
+      clock = clock ;
+      transport = Transport.create ?rng ?nameserver stack
+    }
+
+  let nameserver { transport; _ } = Transport.nameserver transport
+
+  let (>>=) = Transport.bind
+
+  (* result-bind *)
+  let (>>|) a b =
+    a >>= function
+    | Ok a' -> b a'
+    | Error e -> Transport.lift (Error e)
+
+  (* result-bind-and-lift *)
+  let (>>|=) a f = a >>| fun b -> Transport.lift (f b)
+
+  let getaddrinfo (type requested) t ?nameserver (query_type:requested Dns.Rr_map.key) name
+    : (requested, [> `Msg of string]) result Transport.io =
+    let ts = t.clock () in
+    let domain_name = (Domain_name.raw name) in
+
+    match Dns_cache.get t.cache ts domain_name query_type with
+    | Ok (`Entry B (query_type', value), cache) ->
+      t.cache <- cache;
+      assert (query_type' = Obj.magic query_type);
+      Transport.lift @@ Ok(Obj.magic value)
+    | Ok _
+    | Error _ ->
+      let proto, _ = match nameserver with None -> Transport.nameserver t.transport | Some x -> x in
+      let tx, state =
+        Pure.make_query (Transport.rng t.transport)
+          (match proto with `UDP -> `Udp | `TCP -> `Tcp) name query_type
+      in
+      Transport.connect ?nameserver t.transport >>| fun socket ->
+      Logs.debug (fun m -> m "Connected to NS.");
+      (Transport.send socket tx >>| fun () ->
+      Logs.debug (fun m -> m "Receiving from NS");
+      let rec recv_loop acc =
+        Transport.recv socket >>| fun recv_buffer ->
+        Logs.debug (fun m -> m "Read @[<v>%d bytes@]"
+                        (Cstruct.len recv_buffer)) ;
+        let buf = Cstruct.append acc recv_buffer in
```

Comment:
todo: don't do that if `acc` is empty? (depending on whether we care abou tkeeping the reference)